File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/base/ValueShow.java
package wxm.KeepAccount.ui.data.show.note.base;


import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.support.constraint.ConstraintLayout;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.math.BigDecimal;
import java.util.Locale;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import wxm.androidutil.util.UtilFun;
import wxm.KeepAccount.R;


/**
 * use this help show data
 * @author      wxm
 * @version     create：2017/03/28
 */
public class ValueShow extends ConstraintLayout {
    public final static String ATTR_PAY_COUNT = "pay_count";
    public final static String ATTR_PAY_AMOUNT = "pay_amount";
    public final static String ATTR_INCOME_COUNT = "income_count";
    public final static String ATTR_INCOME_AMOUNT = "income_amount";
    private final static String LOG_TAG = "ValueShow";
    @BindView(R.id.tv_pay_tag)
    TextView mTVPayTag;

    @BindView(R.id.tv_pay_count)
    TextView mTVPayCount;

    @BindView(R.id.tv_pay_amount)
    TextView mTVPayAmount;

    @BindView(R.id.tv_income_tag)
    TextView mTVIncomeTag;

    @BindView(R.id.tv_income_count)
    TextView mTVIncomeCount;

    @BindView(R.id.tv_income_amount)
    TextView mTVIncomeAmount;

    @BindView(R.id.tv_balance_amount)
    TextView mTVBalanceAmount;

    @BindView(R.id.iv_pay_line)
    ImageView mIVPayLine;

    @BindView(R.id.iv_income_line)
    ImageView mIVIncomeLine;

    /**
     * can set attributes
     */
    private String mAttrPayCount;
    private String mAttrPayAmount;

    private String mAttrIncomeCount;
    private String mAttrIncomeAmount;

    private int mAttrDMLineLen;

    public ValueShow(Context context, AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(context).inflate(R.layout.vw_value_show, this);
        ButterKnife.bind(this);
        initCompent(context, attrs);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }


    /**
     * adjust attributes
     * @param attrs     new attributes
     */
    public void adjustAttribute(Map<String, Object> attrs) {
        for (String k : attrs.keySet()) {
            switch (k) {
                case ATTR_PAY_COUNT:
                    mAttrPayCount = (String) attrs.get(k);
                    break;

                case ATTR_PAY_AMOUNT:
                    mAttrPayAmount = (String) attrs.get(k);
                    break;

                case ATTR_INCOME_COUNT:
                    mAttrIncomeCount = (String) attrs.get(k);
                    break;

                case ATTR_INCOME_AMOUNT:
                    mAttrIncomeAmount = (String) attrs.get(k);
                    break;
            }
        }

        updateShow();
    }

    /**
     * use date init UI
     * @param context   context for UI
     * @param attrs     data for UI
     */
    private void initCompent(Context context, AttributeSet attrs) {
        // for parameter
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ValueShow);
        try {
            mAttrPayCount = array.getString(R.styleable.ValueShow_szPayCount);
            mAttrPayCount = UtilFun.StringIsNullOrEmpty(mAttrPayCount) ? "0" : mAttrPayCount;

            mAttrPayAmount = array.getString(R.styleable.ValueShow_szPayAmount);
            mAttrPayAmount = UtilFun.StringIsNullOrEmpty(mAttrPayAmount) ? "0" : mAttrPayAmount;

            mAttrIncomeCount = array.getString(R.styleable.ValueShow_szIncomeCount);
            mAttrIncomeCount = UtilFun.StringIsNullOrEmpty(mAttrIncomeCount) ? "0" : mAttrIncomeCount;

            mAttrIncomeAmount = array.getString(R.styleable.ValueShow_szIncomeAmount);
            mAttrIncomeAmount = UtilFun.StringIsNullOrEmpty(mAttrIncomeAmount) ? "0" : mAttrIncomeAmount;

            float dst = getContext().getResources().getDisplayMetrics().density;
            mAttrDMLineLen = array.getLayoutDimension(R.styleable.ValueShow_pxLineLen,
                    (int) (200 * dst));

            updateShow();
        } catch (Exception ex) {
            Log.e(LOG_TAG, "catch ex : " + ex.toString());
        } finally {
            array.recycle();
        }
    }

    /**
     * update UI
     */
    private void updateShow() {
        mTVPayCount.setText(mAttrPayCount);
        mTVPayAmount.setText(mAttrPayAmount);

        mTVIncomeCount.setText(mAttrIncomeCount);
        mTVIncomeAmount.setText(mAttrIncomeAmount);

        BigDecimal bd_pay = new BigDecimal(mAttrPayAmount);
        BigDecimal bd_income = new BigDecimal(mAttrIncomeAmount);
        BigDecimal bd_big = new BigDecimal(Math.max(bd_pay.floatValue(), bd_income.floatValue()));

        BigDecimal bd_balance = bd_income.subtract(bd_pay);
        String sz_b = String.format(Locale.CHINA, "%.02f", bd_balance.floatValue());
        mTVBalanceAmount.setText(sz_b);

        boolean b_p = mAttrPayCount.equals("0");
        mIVPayLine.setVisibility(b_p ? View.GONE : View.VISIBLE);
        mTVPayTag.setVisibility(b_p ? View.GONE : View.VISIBLE);
        mTVPayCount.setVisibility(b_p ? View.GONE : View.VISIBLE);
        mTVPayAmount.setVisibility(b_p ? View.GONE : View.VISIBLE);
        float mPayLinePercent = b_p ? 0 : bd_pay.floatValue() / bd_big.floatValue();

        if (0.01 < mPayLinePercent)
            adjustLineLen(mIVPayLine, mPayLinePercent);

        boolean b_i = mAttrIncomeCount.equals("0");
        mIVIncomeLine.setVisibility(b_i ? View.GONE : View.VISIBLE);
        mTVIncomeTag.setVisibility(b_i ? View.GONE : View.VISIBLE);
        mTVIncomeCount.setVisibility(b_i ? View.GONE : View.VISIBLE);
        mTVIncomeAmount.setVisibility(b_i ? View.GONE : View.VISIBLE);
        float mIncomeLinePercent = b_i ? 0 : bd_income.floatValue() / bd_big.floatValue();

        if (0.01 < mIncomeLinePercent)
            adjustLineLen(mIVIncomeLine, mIncomeLinePercent);

        invalidate();
        requestLayout();
    }

    /**
     * adjust line length
     * @param iv        line need adjust
     * @param percent   new length(percentage)
     */
    private void adjustLineLen(ImageView iv, float percent) {
        ViewGroup.LayoutParams lp = iv.getLayoutParams();
        lp.width = (int) (mAttrDMLineLen * percent);
        iv.setLayoutParams(lp);
    }
}

File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/base/ValueShow.kt
package wxm.KeepAccount.ui.data.show.note.base


import android.content.Context
import android.support.constraint.ConstraintLayout
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView

import java.math.BigDecimal
import java.util.Locale

import kotterknife.bindView
import wxm.androidutil.util.UtilFun
import wxm.KeepAccount.R


/**
 * use this help show data
 * @author      wxm
 * @version     create：2017/03/28
 */
class ValueShow(context: Context, attrs: AttributeSet)
        : ConstraintLayout(context, attrs) {
    private val mTVPayTag: TextView by bindView(R.id.tv_pay_tag)
    private val mTVPayCount: TextView by bindView(R.id.tv_pay_count)
    private val mTVPayAmount: TextView by bindView(R.id.tv_pay_amount)
    private val mTVIncomeTag: TextView by bindView(R.id.tv_income_tag)
    private val mTVIncomeCount: TextView by bindView(R.id.tv_income_count)
    private val mTVIncomeAmount: TextView by bindView(R.id.tv_income_amount)
    private val mTVBalanceAmount: TextView by bindView(R.id.tv_balance_amount)
    private val mIVPayLine: ImageView by bindView(R.id.iv_pay_line)
    private val mIVIncomeLine: ImageView by bindView(R.id.iv_income_line)

    /**
     * can set attributes
     */
    private var mAttrPayCount: String? = null
    private var mAttrPayAmount: String? = null

    private var mAttrIncomeCount: String? = null
    private var mAttrIncomeAmount: String? = null

    private var mAttrDMLineLen: Int = 0

    init {
        LayoutInflater.from(context).inflate(R.layout.vw_value_show, this)
        initComponent(context, attrs)
    }

    /**
     * adjust attributes
     * @param attrs     new attributes
     */
    fun adjustAttribute(attrs: Map<String, Any>) {
        for (k in attrs.keys) {
            when (k) {
                ATTR_PAY_COUNT -> mAttrPayCount = attrs[k] as String
                ATTR_PAY_AMOUNT -> mAttrPayAmount = attrs[k] as String
                ATTR_INCOME_COUNT -> mAttrIncomeCount = attrs[k] as String
                ATTR_INCOME_AMOUNT -> mAttrIncomeAmount = attrs[k] as String
            }
        }

        updateShow()
    }

    /**
     * use date init UI
     * @param context   context for UI
     * @param attrs     data for UI
     */
    private fun initComponent(context: Context, attrs: AttributeSet) {
        // for parameter
        val array = context.obtainStyledAttributes(attrs, R.styleable.ValueShow)
        try {
            mAttrPayCount = array.getString(R.styleable.ValueShow_szPayCount)
            mAttrPayCount = if (UtilFun.StringIsNullOrEmpty(mAttrPayCount)) "0" else mAttrPayCount

            mAttrPayAmount = array.getString(R.styleable.ValueShow_szPayAmount)
            mAttrPayAmount = if (UtilFun.StringIsNullOrEmpty(mAttrPayAmount)) "0" else mAttrPayAmount

            mAttrIncomeCount = array.getString(R.styleable.ValueShow_szIncomeCount)
            mAttrIncomeCount = if (UtilFun.StringIsNullOrEmpty(mAttrIncomeCount)) "0" else mAttrIncomeCount

            mAttrIncomeAmount = array.getString(R.styleable.ValueShow_szIncomeAmount)
            mAttrIncomeAmount = if (UtilFun.StringIsNullOrEmpty(mAttrIncomeAmount)) "0" else mAttrIncomeAmount

            val dst = getContext().resources.displayMetrics.density
            mAttrDMLineLen = array.getLayoutDimension(R.styleable.ValueShow_pxLineLen,
                    (200 * dst).toInt())

            updateShow()
        } catch (ex: Exception) {
            Log.e(LOG_TAG, "catch ex : " + ex.toString())
        } finally {
            array.recycle()
        }
    }

    /**
     * update UI
     */
    private fun updateShow() {
        mTVPayCount.text = mAttrPayCount
        mTVPayAmount.text = mAttrPayAmount

        mTVIncomeCount.text = mAttrIncomeCount
        mTVIncomeAmount.text = mAttrIncomeAmount

        val bdPay = BigDecimal(mAttrPayAmount)
        val bdIncome = BigDecimal(mAttrIncomeAmount)
        val bdBig = BigDecimal(Math.max(bdPay.toFloat(), bdIncome.toFloat()).toDouble())

        val bdBalance = bdIncome.subtract(bdPay)
        val szBalance = String.format(Locale.CHINA, "%.02f", bdBalance.toFloat())
        mTVBalanceAmount.text = szBalance

        val bHavePay = mAttrPayCount == "0"
        mIVPayLine.visibility = if (bHavePay) View.GONE else View.VISIBLE
        mTVPayTag.visibility = if (bHavePay) View.GONE else View.VISIBLE
        mTVPayCount.visibility = if (bHavePay) View.GONE else View.VISIBLE
        mTVPayAmount.visibility = if (bHavePay) View.GONE else View.VISIBLE
        val mPayLinePercent = if (bHavePay) 0.toFloat() else bdPay.toFloat() / bdBig.toFloat()

        if (0.01 < mPayLinePercent)
            adjustLineLen(mIVPayLine, mPayLinePercent)

        val bHaveIncome = mAttrIncomeCount == "0"
        mIVIncomeLine.visibility = if (bHaveIncome) View.GONE else View.VISIBLE
        mTVIncomeTag.visibility = if (bHaveIncome) View.GONE else View.VISIBLE
        mTVIncomeCount.visibility = if (bHaveIncome) View.GONE else View.VISIBLE
        mTVIncomeAmount.visibility = if (bHaveIncome) View.GONE else View.VISIBLE
        val mIncomeLinePercent = if (bHaveIncome) 0.toFloat() else bdIncome.toFloat() / bdBig.toFloat()

        if (0.01 < mIncomeLinePercent)
            adjustLineLen(mIVIncomeLine, mIncomeLinePercent)

        invalidate()
        requestLayout()
    }

    /**
     * adjust line length
     * @param iv        line need adjust
     * @param percent   new length(percentage)
     */
    private fun adjustLineLen(iv: ImageView, percent: Float) {
        val lp = iv.layoutParams
        lp.width = (mAttrDMLineLen * percent).toInt()
        iv.layoutParams = lp
    }

    companion object {
        const val ATTR_PAY_COUNT = "pay_count"
        const val ATTR_PAY_AMOUNT = "pay_amount"
        const val ATTR_INCOME_COUNT = "income_count"
        const val ATTR_INCOME_AMOUNT = "income_amount"
        private val LOG_TAG = ::ValueShow.javaClass.simpleName
    }
}
